File: library/src/main/java/com/chuckerteam/chucker/api/internal/support/ChuckerCrashHandler.java
package com.chuckerteam.chucker.api.internal.support;

import com.chuckerteam.chucker.api.ChuckerCollector;

/**
 * @author Olivier Perez
 */
public class ChuckerCrashHandler implements Thread.UncaughtExceptionHandler {

    private final ChuckerCollector collector;
    private final Thread.UncaughtExceptionHandler defaultHandler;

    public ChuckerCrashHandler(ChuckerCollector collector) {
        this.collector = collector;
        this.defaultHandler = Thread.getDefaultUncaughtExceptionHandler();
    }

    @Override
    public void uncaughtException(Thread t, Throwable e) {
        collector.onError("Error caught on " + t.getName() + " thread", e);
        defaultHandler.uncaughtException(t, e);
    }
}

File: library/src/main/java/com/chuckerteam/chucker/internal/support/ChuckerCrashHandler.kt
package com.chuckerteam.chucker.internal.support

import com.chuckerteam.chucker.api.ChuckerCollector

internal class ChuckerCrashHandler(private val collector: ChuckerCollector) : Thread.UncaughtExceptionHandler {

    private val defaultHandler: Thread.UncaughtExceptionHandler? = Thread.getDefaultUncaughtExceptionHandler()

    override fun uncaughtException(thread: Thread, throwable: Throwable) {
        collector.onError("Error caught on ${thread.name} thread", throwable)
        defaultHandler?.uncaughtException(thread, throwable)
    }
}
